### PR TITLE
feat: added swapExact...SupportingFeeOnTransferTokens() support

### LINKED ISSUE #40: Add support for swapExactTokensForTokensSupportingFeeOnTransferTokens
There is  newer function for tokens that have Fee on Transfer that is causing issues on some transfers, could you please take a look to have a sanctity check [tokenin or tokenout] have this function in the abi to use the correct transfer method to prevent failed TX's

https://uniswap.org/docs/v2/smart-contracts/router02/#swapexacttokensfortokenssupportingfeeontransfertokens
function swapExactTokensForTokensSupportingFeeOnTransferTokens(
  uint amountIn,
  uint amountOutMin,
  address[] calldata path,
  address to,
  uint deadline
) external;

Thanks for keeping this project alive and well!

### PR DESCRIPTION
**Pull request by liquid-8**
Closes #40 #130

### CODE REVIEW

**Comment by liquid-8** [comment]
Comments kinda trivial but whatever.
Router contract has according methods for exact input only so there is no need to rework make_trade_output() as it wraps exact output methods.
If we need these calls for swap then usage gonna be like
`make_trade(token0, token1, qty, ... fee_on_transfer=True)`
If we don't then just omit `fee_on_transfer` param.
Also, please note that univ3 doesn't and won't support this kind of tokens.

File: uniswap/uniswap.py [hunk 1]
```diff
@@ -459,15 +461,26 @@ def _eth_to_token_swap_input(
...
+            else:
+                return self._build_and_send_tx(
+                    self.router.functions.swapExactETHForTokens(
+                        amount_out_min,
+                        [self.get_weth_address(), output_token],
+                        recipient,
+                        self._deadline(),
+                    ),
+                    self._get_tx_params(qty),
+                )
```
Comment (ErikBjare) [inline uniswap/uniswap.py:483]: Could be shorter:

```suggestion
            if fee_on_transfer:
                func = self.router.functions.swapExactETHForTokensSupportingFeeOnTransferTokens
            else:
                func = self.router.functions.swapExactETHForTokens
                
            return self._build_and_send_tx(
                func(
                    amount_out_min,
                    [self.get_weth_address(), output_token],
                    recipient,
                    self._deadline(),
                ),
                self._get_tx_params(qty),
            )
```

File: uniswap/uniswap.py [hunk 2]
```diff
@@ -504,15 +518,26 @@ def _token_to_eth_swap_input(
...
+            else:
+                return self._build_and_send_tx(
+                    self.router.functions.swapExactTokensForETH(
+                        qty,
+                        amount_out_min,
+                        [input_token, self.get_weth_address()],
+                        recipient,
+                        self._deadline(),
+                    ),
+                )
```
Comment (ErikBjare) [inline uniswap/uniswap.py:540]:
```suggestion
            if fee_on_transfer:
            	func = self.router.functions.swapExactTokensForETHSupportingFeeOnTransferTokens
            else:
            	func = self.router.functions.swapExactTokensForETH
            	
            return self._build_and_send_tx(
                func(
                    qty,
                    amount_out_min,
                    [input_token, self.get_weth_address()],
                    recipient,
                    self._deadline(),
                ),
            )
```

File: uniswap/uniswap.py [hunk 3]
```diff
@@ -558,15 +584,26 @@ def _token_to_token_swap_input(
...
+            else:
+                return self._build_and_send_tx(
+                        self.router.functions.swapExactTokensForTokens(
+                        qty,
+                        min_tokens_bought,
+                        [input_token, self.get_weth_address(), output_token],
+                        recipient,
+                        self._deadline(),
+                    ),
+                )
```
Comment (ErikBjare) [inline uniswap/uniswap.py:606]:
```suggestion
            if fee_on_transfer:
            	func = self.router.functions.swapExactTokensForTokensSupportingFeeOnTransferTokens
            else:
            	func = self.router.functions.swapExactTokensForTokens
            	
            return self._build_and_send_tx(
                func(
                    qty,
                    min_tokens_bought,
                    [input_token, self.get_weth_address(), output_token],
                    recipient,
                    self._deadline(),
                ),
            )
```

File: uniswap/uniswap.py [hunk 4]
```diff
@@ -558,15 +584,26 @@ def _token_to_token_swap_input(
...
         elif self.version == 3:
```
Comment (ErikBjare) [inline uniswap/uniswap.py:607]:
```suggestion
        elif self.version == 3:
        	if fee_on_transfer:
        		raise Exception('fee on transfer not supported by Uniswap v3')
```

File: uniswap/uniswap.py [hunk 5]
```diff
@@ -504,15 +518,26 @@ def _token_to_eth_swap_input(
...
+                return self._build_and_send_tx(
+                    self.router.functions.swapExactTokensForETH(
+                        qty,
+                        amount_out_min,
+                        [input_token, self.get_weth_address()],
+                        recipient,
+                        self._deadline(),
+                    ),
+                )
         elif self.version == 3:
```
Comment (ErikBjare) [inline uniswap/uniswap.py:541]:
```suggestion
        elif self.version == 3:
            if fee_on_transfer:
        		raise Exception('fee on transfer not supported by Uniswap v3')
```

File: uniswap/uniswap.py [hunk 6]
```diff
@@ -459,15 +461,26 @@ def _eth_to_token_swap_input(
...
+                return self._build_and_send_tx(
+                    self.router.functions.swapExactETHForTokens(
+                        amount_out_min,
+                        [self.get_weth_address(), output_token],
+                        recipient,
+                        self._deadline(),
+                    ),
+                    self._get_tx_params(qty),
+                )
         elif self.version == 3:
```
Comment (ErikBjare) [inline uniswap/uniswap.py:484]:
```suggestion
        elif self.version == 3:
           	if fee_on_transfer:
                raise Exception('fee on transfer not supported by Uniswap v3')
```

**Review by ErikBjare — COMMENTED** [review]
Nice work! A few comments.

And also: Huge thanks for helping with replying to issues, much appreciated!

**Comment by ErikBjare** [comment]
Do the tests pass for you locally? (CI is not working right for PRs... #88)

**Comment by liquid-8** [comment]
I like your suggestions about method selection but can't commit them as they are marked "outdated" dunno why. Tests are OK.

**Comment by ErikBjare** [comment]
Nice, just run `black` on the files to apply correct formatting, then I'll merge :slightly_smiling_face:

File: uniswap/uniswap.py [hunk 7]
```diff
@@ -568,6 +592,12 @@ def _token_to_token_swap_input(
                 ),
             )
         elif self.version == 3:
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
```
Comment (ErikBjare) [inline uniswap/uniswap.py:600]:
```suggestion
            if fee_on_transfer:
                raise Exception('fee on transfer not supported by Uniswap v3')
```

File: uniswap/uniswap.py [hunk 8]
```diff
@@ -469,6 +475,10 @@ def _eth_to_token_swap_input(
                 self._get_tx_params(qty),
             )
         elif self.version == 3:
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
```
Comment (ErikBjare) [inline uniswap/uniswap.py:481]:
```suggestion
            if fee_on_transfer:
                raise Exception('fee on transfer not supported by Uniswap v3')
```

**Review by ErikBjare — COMMENTED** [review]
Not sure what happened, but some lines were duplicated.

File: uniswap/uniswap.py [hunk 9]
```diff
@@ -514,6 +527,10 @@ def _token_to_eth_swap_input(
                 ),
             )
         elif self.version == 3:
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
+            if fee_on_transfer:
+                raise Exception('fee on transfer not supported by Uniswap v3')
```
Comment (ErikBjare) [inline uniswap/uniswap.py:533]:
```suggestion
```

**Comment by liquid-8** [comment]
> 
> 
> Nice! Just one more thing: run `black` on the files to apply correct formatting.
> 
> After that I'll merge 🙂

Bruh I swear I didn't got ya🤷‍♀️

**Comment by liquid-8** [comment]
> 
> 
> Nice! **Just one more thing: run `black` on the files to apply correct formatting**.
> 
> After that I'll merge 🙂

+++
Seems OK

**Comment by debaenenicolas** [comment]
Hello, 
First thanks for the great repo you built.

I tried to buy reflection tokens with make_trade(token0, token1, qty, ... fee_on_transfer=True) but I still get an error:
"reason : transaction failed"

Do you have an idea why?

Thanks

**Comment by liquid-8** [comment]
>  still get an error:
> "reason : transaction failed"
> Do you have an idea why?

Well, this PR han't been merged yet.

**Comment by ErikBjare** [comment]
LGTM, merging.

Thanks for your contributions @liquid-8!

**Comment by liquid-8** [comment]
> 
> 
> Hello,
> First thanks for the great repo you built.
> 
> I tried to buy reflection tokens with make_trade(token0, token1, qty, ... fee_on_transfer=True) but I still get an error:
> "reason : transaction failed"
> 
> Do you have an idea why?
> 
> Thanks

@debaenenicolas if problem still persists please create thread in Discussions.

**Pull request merged**